File: lc/c/006.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode dummy = { .next = head };
    struct ListNode *fast = &dummy, *slow = &dummy;
    for (int i = 0; i < n; ++i)
        fast = fast->next;

    while (fast->next) {
        fast = fast->next;
        slow = slow->next;
    }

    slow->next = slow->next->next;
    return dummy.next;
}

File: lc/py/006.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0, head)
        fast, slow = dummy, dummy
        for _ in range(n):
            fast = fast.next

        while fast.next:
            fast, slow = fast.next, slow.next

        slow.next = slow.next.next
        return dummy.next
